aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,101 @@
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    /**
+    * TODO: implement per description
+    */
+
+    int index;
+    for(index = 0;index < AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;index++) {
+        if(buffer->entry[(buffer->out_offs + index) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED].buffptr != NULL) {
+            if(char_offset < buffer->entry[(buffer->out_offs + index) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED].size) {
+                *entry_offset_byte_rtn = char_offset;
+                return &buffer->entry[(buffer->out_offs + index) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED];
+            }
+            else {
+                char_offset -= buffer->entry[(buffer->out_offs + index) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED].size;
+            }
+        }
+        else {
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+/**
+* Adds entry @param add_entry to @param buffer in the location specified in buffer->in_offs.
+* If the buffer was already full, overwrites the oldest entry and advances buffer->out_offs to the
+* new start location.
+* Any necessary locking must be handled by the caller
+* Any memory referenced in @param add_entry must be allocated by and/or must have a lifetime managed by the caller.
+*/
+void aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
+{
+    /**
+    * TODO: implement per description
+    */
+    buffer->entry[buffer->in_offs] = *add_entry;
+
+    if(buffer->full) {
+        if(buffer->in_offs == buffer->out_offs) {
+            buffer->out_offs = (buffer->out_offs + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+        }
+        else {
+            if (buffer->out_offs - buffer->in_offs != 1) {
+                buffer->full = false;
+            }
+        }
+
+    }
+    else {
+        if(buffer->out_offs > buffer->in_offs) {
+            if(buffer->out_offs - buffer->in_offs == 1) {
+                buffer->full = true;
+            }
+        }
+        else {
+            if(buffer->in_offs - buffer->out_offs == AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1) {
+                buffer->full = true;
+            }
+        }
+    }
+    buffer->in_offs = (buffer->in_offs + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
```

Comment:
@ashwinhr186 consider modifying your logic to increment in_offs and then check for == to determine full, I think this will be more simple to implement and understand.